ultra/ultra/utils/episode.py

```diff
@@ -219,10 +230,18 @@ def record_step(self, agent_id, infos, rewards, total_step=0, loss_output=None):
         self.steps += 1
         self.agents_itr[agent_id] += 1
 
-    def record_episode(self):
+    def record_episode(self, old_episode=None, eval_rate=None):
         for _, agent_info in self.info[self.active_tag].items():
             agent_info.normalize()
 
+        if (old_episode is not None) and (eval_rate is not None):
+            for agent_id, agent_info in self.info[self.active_tag].items():
+                for key in agent_info.data:
+                    agent_info.data[key] = (
+                        agent_info.data[key] * eval_rate
+                        + old_episode.info[self.active_tag][agent_id].data[key]
+                    ) / eval_rate
+
```

Comment:
Wait sorry could you explain what this is here?

Reply:
Yeah, it's a bit confusing. I want to average training performance over the last eval_rate (lets call it N) episodes. To do this incrementally (one episode at a time), there is a recursive formula: avg_1:(n+1) = (avg_1:n * n + new_data)/(n+1). Since I know the average is going to be over N episodes, I just divide by N = eval_rate instead of n and n+1.

Reply:
Oh right okay gotcha, makes sense. And then these averages will be recorded to TensorBoard every eval_rate episodes: https://github.com/huawei-noah/SMARTS/blob/360e0ebdfa7c90967b4fe9f6f9e3c49a10a334c5/ultra/ultra/train.py#L122-L123 Am I understanding this correctly?

Reply:
Yes exactly